[SHK-65] Feed 조회(상품 기준) 기능 구현

Kim-Changgyu opened the pull request:
### ⛏ 작업 사항
- FeedCustomRepository(QueryDSL) findALlByProduct 메서드 구현
- 피드 상품 기준 조회 기능 관련 단위 및 통합테스트
- 도메인 전역 Bulk Insert 적용 (삽입 쿼리 N개 -> 1개)

### 📝 작업 요약
- 피드 조회(상품 기준) 기능 개발 및 테스트

### 💡 관련 이슈
- Resolved : [SHK-167]


[SHK-167]: https://shoekream.atlassian.net/browse/SHK-167?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Kim-Changgyu committed:
Fix : 피드 등록/수정 시 INSERT 쿼리 N+1 이슈 개선

- JDBC Batch Update (Bulk Insert)를 정의하여 개선
- DataSource URL에 다음과 같은 옵션을 추가하면 하나의 쿼리만 확인 가능
  - rewriteBatchedStatements=true&profileSQL=true&logger=Slf4JLogger&maxQuerySizeToLog=999999

Kim-Changgyu committed:
Feat : [SHK-167] 피드 조회(상품 기준) 기능 구현

Kim-Changgyu committed:
Test : 피드 조회(상품 기준) 기능 단위/통합 테스트

ttaehee commented on src/main/java/com/prgrms/kream/domain/style/repository/FeedProductCustomRepositoryImpl.java:
```diff
@@ -0,0 +1,58 @@
+								new BatchPreparedStatementSetter() {
+									@Override
+									public void setValues(PreparedStatement ps, int i) throws SQLException {
+										Timestamp createdAt =
+												(feedProducts.get(i).getCreatedAt() == null) ?
+														Timestamp.valueOf(LocalDateTime.now()) :
+														Timestamp.valueOf(feedProducts.get(i).getCreatedAt());
```
저도 이번에 JdbcTemplate을 사용했어요
저의 로직과 비교해볼 수 있어 좋네요!

개인적으로 저는 삼항연산자를 가독성면에서 좋아하지 않아, null인 경우만 if문을 타게 구현하였습니다!

Kim-Changgyu replied:
사실 이 Bulk Insert는 저장인 경우에만 한정했기 때문에 생성, 수정 시점에 모두 동일할 것이라고 생각이 들었습니다!
따라서
```java
Timestamp NOW = Timestamp.valueOf(LocalDateTime.now());
```
로 빼둔 다음에 넣어주는 게 로직을 덜 탈 수 있게 해줄 것 같더라구요 !

ttaehee commented on src/main/java/com/prgrms/kream/domain/style/repository/FeedProductCustomRepositoryImpl.java:
```diff
@@ -0,0 +1,58 @@
+												(feedProducts.get(i).getCreatedAt() == null) ?
+														Timestamp.valueOf(LocalDateTime.now()) :
+														Timestamp.valueOf(feedProducts.get(i).getCreatedAt());
+										Timestamp updatedAt =
+												(feedProducts.get(i).getUpdatedAt() == null) ?
+														Timestamp.valueOf(LocalDateTime.now()) :
+														Timestamp.valueOf(feedProducts.get(i).getUpdatedAt());
```
`updatedAt` 부분에서는 null 체크가 불필요할 것 같아요
언제나 현재 시간을 넣으면 되니까요!

ttaehee commented on src/test/java/com/prgrms/kream/domain/style/service/StyleServiceTest.java:
```diff
@@ -209,6 +214,28 @@ void testGetFeedsByMember() {
+				1L,
+				getFeedServiceRequest().cursorId(),
+				getFeedServiceRequest().pageSize());
+		assertThat(getFeedServiceResponses.getFeedServiceResponses()).isEmpty();
```
이 코드는 무엇일까요?!

Kim-Changgyu replied:
상품 식별자로 검색해 온 피드가 존재하지 않음을 증명하기 위한 코드입니다!

ttaehee commented on src/main/java/com/prgrms/kream/domain/style/repository/FeedCustomRepositoryImpl.java:
```diff
@@ -55,6 +58,22 @@ public List<Feed> findAllByMember(Long memberId, Long cursorId, int pageSize) {
+								.select(qFeedProduct.feedId)
+								.from(qFeedProduct)
+								.where(qFeedProduct.productId.eq(productId))),
+						loeFeedId(cursorId)
```
`lt`아니고 `loe` 맞나요?!

Kim-Changgyu replied:
넵! 클라이언트에게 반환되는 lastId(cursor) 값은 다음에 조회할 식별자 번호가 되도록 설계했기 때문에
다음에 조회할 때 해당 식별자부터 결과를 반환해주려고 했습니다!

ttaehee reviewed (approved):
창규님 고생하셨습니다 😀

BnDC commented on src/test/java/com/prgrms/kream/domain/style/service/StyleServiceTest.java:
```diff
@@ -131,9 +135,9 @@ void testUpdate() {
+		verify(feedTagRepository).batchUpdate(Mockito.<FeedTag>anyList());
 		verify(feedProductRepository).deleteAllByFeedId(FEED.getId());
-		verify(feedProductRepository).saveAll(Mockito.<FeedProduct>anyIterable());
+		verify(feedProductRepository).batchUpdate(Mockito.<FeedProduct>anyList());
```
`Itarable` 에서 `List`로 좀 더 구체적인 것으로 변경하셨군요!

BnDC reviewed (approved):
구현하느라 고생하셨습니다.🙂

Kim-Changgyu committed:
Refactor : [SHK-65] 피드백 반영

- CustomRepository 메서드명 통일 (Product 도메인 참조)
- 생성, 수정시점 변수화
- 본문 길이 Validation 추가
